### DISCUSSION TITLE
problem converting a string of chars into a single number in micropython

### DISCUSSION THREAD

**Discussion by theproff** [micropython/micropython#12445] [General]
hi, My name is Tony, im very new to Python.
my problem is im reading data into an esp32 via Uart, the data i get is a string of ascii characters that represent a floating point number.
im trying to convert that string to a physical number to pass onto adafruit via mqtt.

the string consists of: 3735352e3030 this is 755.00 and i cannot get it as a variable, ive tried various ways shown in examples on google but non work
can anyone please help

im using Thonny

this is my code

```py
def get_credit():

    uart_meter.write('cR1\r')                 #send command to device to get response

    while uart_meter.any() == 0:
        strMsg = uart_meter.read()
        print(strMsg)                              # the message it prints is b'755.00' which is correct
        break
```

thanks for any help

**Comment by jimmo** [comment]
@theproff  The string `"3735352e3030"` is the hexadecimal encoding of the string `"755.00"`. So you can decode it using bytes.fromhex, and use `float()` to convert it to a numeric value.

e.g.

```py
strMsg = uart_meter.read()
decoded = bytes.fromhex(strMsg)
value = float(decoded)
print(value)
```

FYI Please see [Welcome (Please read before posting)](https://github.com/orgs/micropython/discussions/9111) for how to insert code snippets in posts.

**Reply by theproff** [reply]
hi Jimmo, thanks for the reply, im getting an error

  File "meter.py", line 23, in get_credit
TypeError: object with buffer protocol required

line 23 is:- decoded = bytes.fromhex(strMsg)

any ideas, thanks and i will read up on sending code snippets

**Reply by robert-hh** [reply]
Maybe `strMsg` is empty and has the value None. uart.read() returns what is present at the time of the call. If there is nothing available, it returns None. So you have to ensure first, that you got the data.

**Reply by theproff** [reply]
ps. if i run this code in the thonny shell it works fine

**Reply by jimmo** [reply]
Try adding `print(strMsg)` above the `decoded = bytes.fromhex(strMsg)` line, then you'll be able to see what's going on.

**Reply by theproff** [reply]
ok Robert, i will check that, ill first just generate a string with hex values to test. then ensure i have data received correctly

**Reply by theproff** [reply]
hi jimmo and robert, thanks for all your help, im sure im getting there

this code works fine if i just generate a hex string but not via uart even though the uart is receiving the right numbers. see below
i put the print before the decode line

Started
b'774.00'                                   this is the printout after it received the data
Traceback (most recent call last):
  File "<stdin>", line 31, in <module>
  File "meter.py", line 26, in get_credit
ValueError: non-hex digit found

line 26 is :- decoded = bytes.fromhex(strMsg)

**Reply by jimmo** [reply]
It looks like strMsg isn't hex encoded then? Rather it's just a string representation of the number. You can just use `float(strMsg)` directly?

Where did you first get the "3735352e3030" string from? Was that actually something you received from the uart, or maybe how a data sheet described the uart data?

**Reply by theproff** [reply]
hi Jimmo, the first string was direct from uart, 755.00. ill try the float direct which im sure ive tried before....
ok, that works fine without the decoded bit...
Started
b'774.00'     #this is the string contents
774.0     #and this is the float(strMsg) result
end

thanks for your help here, ive been on with this for 2 days.......
thanks to Jimmo and Robert appreciate this..
